test/metabase/query_processor/pivot_test.clj

```diff
@@ -15,54 +16,167 @@
+             [0    ]
+             [0 1  ]
+             ;; "grand totals" row
+             [    3]
+             ;; bottom right corner
+             [     ]])
+           (#'pivot/breakout-combinations 4 [0 1 2] [3])))))
+
+(deftest group-number-test
+  (doseq [[indecies expected] {[0]     6
+                               [0 1]   4
+                               [0 1 2] 0
+                               []      7}]
+    (is (= expected
+           (#'pivot/group-number 3 indecies)))))
+
+(deftest validate-pivot-rows-cols-test
+  (testing "Should throw an Exception if you pass in invalid pivot-rows"
+    (is (thrown-with-msg?
+         clojure.lang.ExceptionInfo
+         #"Invalid pivot-rows: specified breakout at index 3, but we only have 3 breakouts"
+         (#'pivot/breakout-combinations 3 [0 1 2 3] []))))
+  (testing "Should throw an Exception if you pass in invalid pivot-cols"
+    (is (thrown-with-msg?
+         clojure.lang.ExceptionInfo
+         #"Invalid pivot-cols: specified breakout at index 3, but we only have 3 breakouts"
+         (#'pivot/breakout-combinations 3 [] [0 1 2 3]))))
+  ;; TODO -- we should require these columns to be distinct as well (I think?)
+  ;; TODO -- require all numbers to be positive
+  ;; TODO -- can you specify something in both pivot-rows and pivot-cols?
+  )
+
+(defn- test-query []
+  (mt/dataset sample-dataset
+    (mt/$ids orders
+      {:database     (mt/id)
+       :type         :query
+       :query        {:source-table $$orders
+                      :aggregation  [[:count]]
+                      :breakout     [$product_id->products.category
+                                     $user_id->people.source
+                                     !year.created_at]
+                      :filter       [:and
+                                     [:= $user_id->people.source "Facebook" "Google"]
+                                     [:= $product_id->products.category "Doohickey" "Gizmo"]
+                                     [:time-interval $created_at -2 :year {}]]}
+       :pivot-rows [0 1 2]
+       :pivot-cols []})))
+
+(deftest allow-snake-case-test
+  (testing "make sure the stuff works with either normal lisp-case keys or snake case. It should only ever see lisp-case"
+    (is (= (mt/rows (pivot/run-pivot-query (test-query)))
+           (mt/rows (pivot/run-pivot-query (set/rename-keys (test-query)
+                                                            {:pivot-rows :pivot_rows, :pivot-cols :pivot_cols})))))))
+
+(deftest dont-return-too-many-rows-test
+  (testing "Make sure pivot queries don't return too many rows (#14329)"
+    (mt/dataset sample-dataset
+      (let [results (pivot/run-pivot-query (test-query))
+            rows    (mt/rows results)]
+        ;; TODO - fix me
+        #_(is (= ["Category"
+                  "Source"
+                  "Created At"
+                  "pivot-grouping"
+                  "Count"]
+                 (map :display_name (mt/cols results))))
```

Comment:
the columns that come back in metadata for the `run-pivot-query` function are wrong... not sure if there's an open issue to fix this or not yet

Reply:
Ahhh that's likely an issue from the reordering of the columns that occurred with fixing the previous pgsql issue. There is no open issue for it.

Reply:
fixed it